Drivers/Inc/CANbus.h

```diff
@@ -18,59 +20,57 @@ typedef enum {
 	CAR_STATE = 0x580,
 	CHARGE_ENABLE = 0x10C,
 	SUPPLEMENTAL_VOLTAGE = 0x10B,
-	STATE_OF_CHARGE = 0x106
+	STATE_OF_CHARGE = 0x106,
+	MOTOR_DRIVE = 0x221,
+	MOTOR_POWER = 0x222,
+	MOTOR_RESET = 0x223,
+	MOTOR_STATUS = 0x241,
+	MOTOR_VELOCITY = 0x243,
+	DO_NOT_USE_FORCE_ENUM_SIZE = 0xFFFFFFFF //force the enum values to be 32 bits wide. This value should not be used. This is here because the BSP layer expects a 32 bit ptr for ID's
 } CANId_t;
 
 
-typedef union {
-	uint8_t b;
-	uint16_t h;
-	uint32_t w;
-	uint64_t d;
-} CANData_t;
 
-typedef struct {
-	uint8_t idx : 8;
-	uint8_t bytes : 8;
-	CANData_t data;
-} CANPayload_t;
 
 /**
- * Data type for message queue
+ * Standard CAN packet
 */
 typedef struct {
-	CANPayload_t payload;
-	CANId_t id;
-}CANMSG_t;
+	CANId_t ID; 		//ID of message
+	uint8_t idx; 		//FOR TRANSMIT ONLY: if message is part of a sequence of messages (for messages longer than 64 bits), this indicates the index of the message. Recieve will not touch this
+	bool idxEn; 		//FOR TRANSMIT ONLY: whether to use idx or not. Recieve will not touch this.
+	uint8_t size; 		//size of this particular message IN BYTES. On writes, this should not Exceed 8 bytes for non-idx messages, and should not exceed 7 for idx messages.
+	uint64_t data; 	//data of the message
+} CANDATA_t;
+
 
 typedef enum {CAN_BLOCKING=0, CAN_NON_BLOCKING} CAN_blocking_t;
 
 /**
- * @brief   Initializes the CAN system
- * @param   None
+ * @brief   Initializes the CAN system for a given bus
+ * @param   bus The bus to initialize. You can either use CAN_1, CAN_3, or the convenience macros CARCAN and MOTORCAN
  * @return  None
  */
-void CANbus_Init(void);
+void CANbus_Init(CAN_t bus);
 
 /**
- * @brief   Transmits data onto the CANbus.
+ * @brief   Transmits data onto the CANbus. Transmits up to 8 bytes at a time. If more is needed, 
  * @param   id : CAN id of the message
  * @param 	payload : the data that will be sent.
  * @param   blocking: Whether or not the Send should be blocking or not
+ * @param   bus: Which bus to transmit on
  * @return  ERROR if data wasn't sent, otherwise it was sent.
  */
-ErrorStatus CANbus_Send(CANId_t id, CANPayload_t payload,CAN_blocking_t blocking);
-
+ErrorStatus CANbus_Send(CANDATA_t CanData,CAN_blocking_t blocking, CAN_t bus);
 
 /**
- * @brief   Checks if the CAN ID matches with Motor disable ID
- * @param   canline (not implemented, default is CAN1) can line to read from
- * @param 	id CAN msg ID
- * @param 	buffer pointer to buffer in which to store the can msg
- * @param   blocking whether or not this Read should be a blocking read or a nonblocking read
- * @return  1 if ID matches and 0 if it doesn't
+ * @brief   Reads a CAN message from the CAN hardware and returns it to the provided pointers. 
+ * 	DOES NOT POPULATE IDXen or IDX. You have to manually inspect the first byte and the ID
```

Comment:
If you need to manually populate ```IDXen``` and ```IDX```, I'm not sure how useful it is to have those fields in the CAN datatype. It might be easier to have a lookup function/table for each ID which CAN reads and writes can reference for enabling the IDX byte.

Reply:
This makes sense, although you'd still need idx so that you could specify the position of the message in the array you're sending. I'll add a LUT that matches messages to idxEN values though.